### PR TITLE
feat: handle multiple identical ids

### PR DESCRIPTION
**Pull request by taavilooke**
Handle the case where there are multiple entries for identical `id`s in the database. Previously the code gave an error in this case and showed this as no card in the front end. Since instructors can add cards themselves now, we need to handle this case.

Added logic to find the highest value card that is currently valid and show that in the front-end.

Also fixed a bug that showed a valid card with no expiry date if no expiry date was set in the database. Now it shows as no card.

### CODE REVIEW

**Comment by taavilooke** [comment]
Found two errors:

1) The PR removes the line that returns `formFillTime` from `findById` function. It seemed that the UI doesn't use it anywhere so there's no reason to hand it over to the UI. However the tests do expect it right now. Let's remove it from the tests as well.
2) Testcase `40000000000` in the current implementation returns a certificate: "none". But in the PR's implementation it returns "error: not found". I think this is a more appropriate response, since the certificate is not correct.

Rauno, what do you think?

**Comment by rauno56** [comment]
To reply to your points:

1. Thing to keep in mind is that our UI is not the only consumer for the API. However with `formFillTime`... it probably is not used anywhere maybe?
2. I think that's fair because we control what's inserted in the registry. It would be useful to know whether one is dealing with invalid data or really-not-fonud tho.

Otherwise looks good!

- [ ] Fix the linting errors (I pushed some rules I noticed your code not following, but rest of the codebase having, so linter will guide you through it)
- [ ] Fix broken tests
- [ ] Add a test that tests the added case. `fixture.txt is` just copy-paste of the test rows in the doc so feel free to add those cases there first and copy the result from there without editing the `fixture.txt` directly.

File: api/_db.data-utils.js [hunk 1]
```diff
@@ -55,6 +55,27 @@ const parseDate = (rawValue) => {
...
 
+const findHighestValidCertificate = (certificates) => {
+	// find the certificate that is not expired and has the highest value (red is higher than green).
+	// nonexpired green has to beat expired red, but red with shorter expiryTime has to beat any green.
+	return certificates.reduce((previous, current) =>{
+		if (!previous) {
+			return current;
+		}
+		if (parseDate(current.expiryTime) > Date.now()){
+			// TODO: maybe we should find the certificate with the longest expiry time
```
Comment (rauno56) [inline api/_db.data-utils.js:66]: I think we should find the highest-degree **valid** certificate.
If multiple found, the last issued valid certificate.
And if that not found, the last issued invalid certificate.

File: api/_db.data-utils.js [hunk 2]
```diff
@@ -55,6 +55,27 @@ const parseDate = (rawValue) => {
...
+			// currently the first valid one is kept
+			if (previous.certificate != CODE.RED && current.certificate == CODE.RED){
+				return current;
+			}
+			if (previous.certificate == CODE.GREEN && current.certificate == CODE.GREEN){
```
Comment (rauno56) [inline api/_db.data-utils.js:71]: This makes an assumption that order of which the data is presented(not "entered" because the process currently does not really assert order) to this function matters. Do we really want to make that assumption?

File: api/_db.data-utils.js [hunk 3]
```diff
@@ -107,28 +128,29 @@ export const findById = (data, id) => {
...
+		};
+
+	// remove certificates without expiryTime set
+	// the front-end does not handle this case, so we handle it here
+	}).filter((row) => row.expiryTime);
+
+	// no valid certificates are found
+	assert(parsedCertificates.length, 'Not found');
+
+	const highestValueCertificate = findHighestValidCertificate(parsedCertificates);
```
Comment (rauno56) [inline api/_db.data-utils.js:153]: This doesn't need assignment, better to directly return.

**Review by rauno56 — COMMENTED** [review]
A solid stab at the problem, few comments in the post and code.

**Comment by taavilooke** [comment]
- Should we rebase the whole branch to add conventioncommits style messages to all the commits?
- Should we clean up the history somehow with the merges? How to do that?

**Comment by rauno56** [comment]
> Should we rebase the whole branch to add conventioncommits style messages to all the commits?<br>Should we clean up the history somehow with the merges? How to do that?

We should squash-merge PRs that include one feature like this one, so the individual commit messages do not matter. Only the PR title which will be used for the commit that lands in the main.

File: api/_db.data-utils.js [hunk 4]
```diff
@@ -53,6 +53,40 @@ const parseDate = (rawValue) => {
...
+	let bestRedCard = null;
+	let bestGreenCard = null;
+	let anyExpiredCard = null;
+	let anyInvalidCard = null;
+
+	for (const card of certificates) {
+		const expiryDate = parseDate(card.expiryTime);
+		const examDate = parseDate(card.examTime);
+
+		if ((card.certificate !== CODE.RED && card.certificate !== CODE.GREEN) || !expiryDate) {
```
Comment (rauno56) [inline api/_db.data-utils.js:72]: `!expiryDate` check semantically belongs to the next `if` and since you're then later checking `card.certificate === CODE.{GREEN,RED}` anyway, I'd replace this whole block with `else` in the end.

File: api/_db.data-utils.js [hunk 5]
```diff
@@ -79,7 +113,7 @@ const formFillTimeHeader = 'formFillTime';
 export const findById = (data, id) => {
 	assertValidId(id);
 
-	// Ignore human-readable headers(the first row), because those can change any time
+	// ignore human-readable headers(the first row), because those can change any time
```
Comment (rauno56) [inline api/_db.data-utils.js:116]: Why this change. It's not consistent in any way.

I usually have `// ` in the beginning of every comment (note the space) and capitalize the first char.

Just make them consistent at minimum.

File: api/_db.data-utils.js [hunk 6]
```diff
@@ -105,28 +139,32 @@ export const findById = (data, id) => {
...
-	}
+		const formFillDate = parseDate(row[formFillTimeColumnIdx]);
+		const expiryDate = parseDate(row[expiryDateColumnIdx]) || getExpiryTimeFromFormFillTime(formFillDate) || null;
+
+		return {
+			id,
+			certificate,
+			name: row[nameColumnIdx],
+			examiner: row[examinerColumnIdx] || null,
+			examTime: formatDate(parseDate(row[examDateColumnIdx])),
```
Comment (rauno56) [inline api/_db.data-utils.js:153]: Let's parse(at the time of loading the info, eg here) and format(at the very end when returning from here) each of the dates once.

File: test/db.js [hunk 7]
```diff
@@ -37,49 +35,54 @@ const testCases = {
...
-		expiryTime: '2024-03-09'
-	},
+	// '70000000000': {
+	// 	id: '70000000000',
+	// 	certificate: 'green',
+	// 	name: 'Ain Vormiga',
+	// 	examiner: 'Eerik Eksamineerija',
+	// 	examTime: null,
+	// 	expiryTime: '2024-03-09'
+	// },
```
Comment (rauno56) [inline test/db.js:63]: If we don't use those comments, we shouldn't keep them here. Git helps us to know the history.

It looks like this example has been deleted, and another one `90000..` was created for the same purpose.

File: api/_db.data-utils.js (hunk 4 above)
Comment (taavilooke) [reply api/_db.data-utils.js:72]: We need to check  for cards with `expiryDate===null` right at the beginning so that cards without it would be filtered out right away: otherwise these would fall into `expiryDate < Date.now()` and we need to differentiate invalid and expired cards. I refactored it a bit so that the logic would be clearer.

File: test/db.js (hunk 7 above)
Comment (taavilooke) [reply test/db.js:63]: Actually 4xx and 7xx were changed to invalidCases, see below. I deleted the comments.

File: api/_db.data-utils.js (hunk 6 above)
Comment (taavilooke) [reply api/_db.data-utils.js:153]: Good idea, this makes everything easier!

**Review by rauno56 — APPROVED** [review]
How are we handling a case when someones expired red will be replaced by valid green?

If that's handled (or knowingly left unhandled) correctly, feel free to squash-merge this.

**Comment by taavilooke** [comment]
Right now it's meant to work so that if the red card is expired the valid
green will be shown. However in practice the red card will always have a
later expiration date than the green card and when the red card expires the
green card will have expired earlier than that. Is this what you meant?
